File: 0021. spiral-traverse.py
# O(n)
def spiralTraverse(arr):
    res = []
    startRow, endRow = 0, len(arr) - 1
    startCol, endCol = 0, len(arr[0]) - 1

    while startRow <= endRow and startCol <= endCol:
        for col in range(startCol, endCol+1):
            res.append(arr[startRow][col])

        for row in range(startRow+1, endRow + 1):
            res.append(arr[row][endCol])

        for col in reversed(range(startCol, endCol)):
            res.append(arr[endRow][col])

        for row in reversed(range(startRow+1, endRow)):
            res.append(arr[row][startCol])

        startRow += 1
        endRow -= 1
        startCol += 1
        endCol -= 1

    return res


if __name__ == "__main__":
    arr = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]]
    for i in range(0, len(arr)):
        for j in range(0, len(arr[i])):
            print(arr[i][j], end=' ')
        print()
    print(spiralTraverse(arr))

File: 0021. spiral-traverse.cpp
#include <iostream>
#include <vector>
using namespace std;

void print(const vector<vector<int>> &arr)
{
    for (int i = 0; i < arr.size(); i++)
    {
        for (int j = 0; j < arr[i].size(); j++)
        {
            cout << arr[i][j] << " ";
        }
        cout << endl;
    }
}

vector<int> spiralTraverse(const vector<vector<int>> &vect)
{
    vector<int> v;
    if (vect.empty())
        return {};
    int startRow = 0;
    int endRow = vect.size() - 1;
    int startCol = 0;
    int endCol = vect[0].size() - 1;
    int dir = 0;
    while (startCol <= endCol and startRow <= endRow)
    {
        switch (dir)
        {
        case 0:
            for (int col = startCol; col <= endCol; col++)
                v.push_back(vect[startRow][col]);
            startRow += 1;
            break;
        case 1:
            for (int row = startRow; row <= endRow; row++)
                v.push_back(vect[row][endCol]);
            endCol -= 1;
            break;
        case 2:
            for (int col = endCol; col >= startCol; col--)
                v.push_back(vect[endRow][col]);
            endRow -= 1;
            break;
        case 3:
            for (int row = endRow; row >= startRow; row--)
                v.push_back(vect[row][startCol]);
            startCol += 1;
            break;
        }
        dir = (dir + 1) % 4;
    }
    return v;
}

int main(int argc, char const *argv[])
{
    vector<vector<int>> vect{{1, 2, 3, 4},
                             {5, 6, 7, 8},
                             {9, 10, 11, 12}};
    print(vect);
    vector<int> v = spiralTraverse(vect);
    for (int i : v)
        cout << i << " ";
    return 0;
}
